docs/dev/guidelines/criteria-builder.rst

```diff
@@ -0,0 +1,294 @@
+            return (root, query, criteriaBuilder) -> {
+                String[] columns = new String[] { User_.LOGIN, User_.EMAIL, User_.FIRST_NAME, User_.LAST_NAME };
+                Predicate[] predicates = Arrays.stream(columns).map((column) -> criteriaBuilder.like(root.get(column), extendedSearchTerm)).toArray(Predicate[]::new);
+
+                return criteriaBuilder.or(predicates);
+            };
+        }
+
+    .. code-block:: sql
+
+        SELECT DISTINCT user FROM jhi_user user
+        WHERE user.login LIKE ?
+            OR user.email LIKE ?
+            OR user.first_name LIKE ?
+            OR user.last_name LIKE ?
+        ORDER BY user.id ASC limit ?
+
+- Specification that matches all selected courses:
+
+    .. code-block:: java
+
+        public static Specification<User> getAllUsersMatchingCourses(Set<Long> courseIds) {
+        	return (root, query, criteriaBuilder) -> {
+            	Root<Course> courseRoot = query.from(Course.class);
+
+                Join<User, String> group = root.join(User_.GROUPS, JoinType.LEFT);
+
+                // Select all possible group types
+                String[] columns = new String[] { Course_.STUDENT_GROUP_NAME, Course_.TEACHING_ASSISTANT_GROUP_NAME, Course_.EDITOR_GROUP_NAME, Course_.INSTRUCTOR_GROUP_NAME };
+                Predicate[] predicates = Arrays.stream(columns).map((column) -> criteriaBuilder.in(courseRoot.get(column)).value(group)).toArray(Predicate[]::new);
+
+                // The course needs to be one of the selected
+                Predicate inCourse = criteriaBuilder.in(courseRoot.get(Course_.ID)).value(courseIds);
+
+                group.on(criteriaBuilder.or(predicates));
+
+                query.groupBy(root.get(User_.ID)).having(criteriaBuilder.equal(criteriaBuilder.count(group), courseIds.size()));          
+
+        		return criteriaBuilder.in(courseRoot.get(Course_.ID)).value(courseIds);
+            }
+        }
+
+    .. code-block:: sql
+
+        SELECT DISTINCT user FROM jhi_user user
+        CROSS JOIN course course
+        LEFT OUTER JOIN user_groups groups ON user.id = groups.user_id
+        AND (course.student_group_name IN ( groups.`groups` )
+            OR course.teaching_assistant_group_name IN ( groups.`groups` )
+            OR course.editor_group_name IN ( groups.`groups` )
+            OR course.instructor_group_name IN ( groups.`groups` )
+        WHERE (user.login LIKE ?
+            OR user.email LIKE ?
+            OR user.first_name LIKE ?
+            OR user.last_name LIKE ?)
+        AND ( course.id IN ( ? ) )
+        GROUP BY user.id
+        HAVING Count(groups.`groups`) = ?
+        ORDER BY user.id ASC
+        LIMIT ?
+
+- Specification to get distinct results:
+
+    .. code-block:: java
+
+        public static Specification<User> distinct() {
+        	return (root, query, criteriaBuilder) -> {
+        		query.distinct(true);
+                return null;
```

Comment:
This is the only place with `return null`. It should probably be mentioned that this is fine when the `criteriaBuilder` is not changed but only the `query` is mutated in-place (Intuitive explanation. Is that even the actual rule why it is okay here?).

Reply:
Yes since null predicates/specifications are ignored :+1: 

```java
static <T> Specification<T> composed(@Nullable Specification<T> lhs, @Nullable Specification<T> rhs,
			Combiner combiner) {
    return (root, query, builder) -> {

	Predicate thisPredicate = toPredicate(lhs, root, query, builder);
	Predicate otherPredicate = toPredicate(rhs, root, query, builder);

	if (thisPredicate == null) {
		return otherPredicate;
	}

	return otherPredicate == null ? thisPredicate : combiner.combine(builder, thisPredicate, otherPredicate);
    };
}

@Nullable
private static <T> Predicate toPredicate(@Nullable Specification<T> specification, Root<T> root, CriteriaQuery<?> query,
	CriteriaBuilder builder) {
    return specification == null ? null : specification.toPredicate(root, query, builder);
}
```